src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Runners/LoggingRunnerAdapter.cs

```diff
@@ -108,42 +108,77 @@ public async Task StartAsync(CancellationToken token)
                 _outputHelper.WriteLine("Already exited.");
+                exitCode = _runner.ExitCode;
             }
             else
             {
                 _outputHelper.WriteLine("Waiting for exit...");
                 await _runner.WaitForExitAsync(token).ConfigureAwait(false);
+                exitCode = _runner.ExitCode;
             }
-            _exitCode = _runner.ExitCode;
-            _outputHelper.WriteLine("Exit Code: {0}", _runner.ExitCode);
-            return _runner.ExitCode;
+            _outputHelper.WriteLine("Exit Code: {0}", _exitCode);
+            _exitCode = exitCode;
+            return exitCode.Value;
+        }
+
+        public async Task ReadToEnd(CancellationToken token)
+        {
+            // First we need to wait for the process to end
+            await WaitForExitAsync(token);
+
+            // Then tell the readers to end
+            _finishReads = true;
+            Task errorWaiter = _standardErrorTask.SafeAwait(_outputHelper);
+            Task stdWaiter = _standardOutputTask.SafeAwait(_outputHelper);
+
+            // Wait for everything (with the cancellation token allowed
+            await Task.Run(() => Task.WaitAll(new Task[] { stdWaiter, errorWaiter }, token)).ConfigureAwait(false);
         }
 
-        private static async Task ReadLinesAsync(StreamReader reader, List<string> lines, Action<string> callback, CancellationToken token)
+        private async Task ReadLinesAsync(StreamReader reader, List<string> lines, Action<string> callback, CancellationToken cancelToken)
         {
             // Closing the reader to cancel the async await will dispose the underlying stream.
-            // Technically, this means the reader/stream cannot be used after cancelling reading of lines
+            // Technically, this means the reader/stream cannot be used after canceling reading of lines
             // from the process, but this is probably okay since the adapter is already logging each line
             // and providing a callback to callers to read each line. It's unlikely the reader/stream will
-            // be accessed after this adapter is diposed.
-            using var _ = token.Register(() => reader.Close());
+            // be accessed after this adapter is disposed.
+            using var cancelReg = cancelToken.Register(() => reader.Close());
 
             try
             {
-                while (true)
+                bool readAborted = false;
+                while (!_finishReads)
                 {
                     // ReadLineAsync does not have cancellation
                     string line = await reader.ReadLineAsync().ConfigureAwait(false);
 
                     if (null == line)
+                    {
+                        readAborted = true;
                         break;
+                    }
 
                     lines.Add(line);
                     callback?.Invoke(line);
                 }
+                if (!readAborted && _finishReads)
```

Comment:
This solves an issue where during shutdown, the rest of the stream can be dropped on the floor. It doesn't show up in the normal tests for `collect` because dotnet-monitor takes a while to shutdown and gives the `.ReadLineAsync()` operations a while to catch up but for `generatekey` it runs too fast and the log gets clipped.
The readAborted is the preserve the initial behavior of successfully cleaning up everything on `.DisposeAsync()`

Reply:
I added some comments explaining what this is doing.